src/Settings/Builders/GhcCabal.hs

```diff
@@ -2,37 +2,73 @@ module Settings.Builders.GhcCabal (
     ghcCabalBuilderArgs
     ) where
 
-import Hadrian.Haskell.Cabal
+import Data.Maybe (fromJust)
 
+import Builder ( ArMode ( Pack ) )
 import Context
 import Flavour
+import GHC.Packages
+import Hadrian.Builder (getBuilderPath, needBuilder )
+import Hadrian.Haskell.Cabal
 import Settings.Builders.Common
 
 ghcCabalBuilderArgs :: Args
-ghcCabalBuilderArgs = builder GhcCabal ? do
+ghcCabalBuilderArgs = mconcat
+  [ builder (GhcCabal Conf) ? do
     verbosity <- expr getVerbosity
     top       <- expr topDirectory
-    path      <- getBuildPath
-    notStage0 ? expr (need inplaceLibCopyTargets)
+    path      <- getContextPath
+    stage     <- getStage
     mconcat [ arg "configure"
-            , arg =<< pkgPath <$> getPackage
+            -- don't strip libraries when cross compiling.
+            -- XXX we need to set --with-strip= (stripCmdPath :: Action FilePath), and if it's ':' disable
+            --     stripping as well. As it is now, I believe we might have issues with stripping on
+            --     windows, as I can't see a consumre of `stripCmdPath`.
+            , crossCompiling ? pure [ "--disable-executable-stripping", "--disable-library-stripping" ]
+            , arg "--cabal-file"
+            , arg =<< fromJust . pkgCabalFile <$> getPackage
+            , arg "--distdir"
             , arg $ top -/- path
+            , arg "--ipid"
+            , arg "$pkg-$version"
+            , arg "--prefix"
+            , arg "${pkgroot}/.."
             , withStaged $ Ghc CompileHs
             , withStaged (GhcPkg Update)
+            , withBuilderArgs (GhcPkg Update stage)
             , bootPackageDatabaseArgs
             , libraryArgs
             , configureArgs
             , bootPackageConstraints
             , withStaged $ Cc CompileC
-            , notStage0 ? with Ld
+            , notStage0 ? with (Ld stage)
             , withStaged (Ar Pack)
             , with Alex
             , with Happy
-            , verbosity < Chatty ? pure [ "-v0", "--configure-option=--quiet"
-                , "--configure-option=--disable-option-checking"  ] ]
+            , verbosity < Chatty ?
+              pure [ "-v0", "--configure-option=--quiet"
+                   , "--configure-option=--disable-option-checking"
+                   ]
+            ]
+  , builder (GhcCabal Copy) ? do
+      mconcat [ arg "copy"
+              , getInputs
+              ]
+  , builder (GhcCabal Reg) ? do
+      top       <- expr topDirectory
+      path      <- getContextPath
+      mconcat [ arg "register"
+              , arg =<< pkgPath <$> getPackage
+              , arg $ top -/- path
+              , stagedBuilderPath (Ghc CompileHs)
+              , stagedBuilderPath (GhcPkg Update)
+              , getInputs
+              ]]
+
 
 -- TODO: Isn't vanilla always built? If yes, some conditions are redundant.
 -- TODO: Need compiler_stage1_CONFIGURE_OPTS += --disable-library-for-ghci?
+-- TODO: should `elem` be `wayUnit`?
```

Comment:
So, should it or not? Presumably only we can answer this question.

Reply:
Fixed this in an upcoming commit.